### PR TITLE
Type hints for tests.federation

### PR DESCRIPTION
**Pull request by DMRobertson**

### CODE REVIEW

File: tests/federation/test_federation_catch_up.py [hunk 1]
```diff
@@ -55,8 +61,13 @@ def default_config(self) -> JsonDict:
         config["federation_sender_instances"] = None
         return config
 
-    async def record_transaction(self, txn, json_cb):
-        if self.is_online:
+    async def record_transaction(
+        self, txn: Transaction, json_cb: Optional[Callable[[], JsonDict]]
+    ) -> JsonDict:
+        if json_cb is None:
+            # The tests seem to expect that this method raises in this situation.
```
Comment (clokep) [inline tests/federation/test_federation_catch_up.py:68]: I guess previously via a `NoneType has no attribute` error or something?

File: tests/federation/test_federation_sender.py [hunk 2]
```diff
@@ -473,13 +481,15 @@ def test_upload_signatures(self):
             self.assertEqual(edu["edu_type"], EduTypes.DEVICE_LIST_UPDATE)
             c = edu["content"]
             if stream_id is not None:
-                self.assertEqual(c["prev_id"], [stream_id])
+                # type-ignore: mypy marks the next line as unreachable; it does not seem
+                # to recognise the assignment to stream_id within the body of this loop
```
Comment (clokep) [inline tests/federation/test_federation_sender.py:485]: I think if you add an explicit `stream_id: Optional[str] = None` you can remove the ignore?

Comment (DMRobertson) [reply tests/federation/test_federation_sender.py:485]: Odd, I could have sworn I did that somewhere earlier in the function... seemingly not!

Comment (DMRobertson) [reply tests/federation/test_federation_sender.py:485]: Ahh, I'm thinking of line 568 below.

Comment (DMRobertson) [reply tests/federation/test_federation_sender.py]: This is 9d770bf37e

**Review by clokep — APPROVED** [review]

**Pull request merged**